tests/test_package_swap.py

```diff
@@ -0,0 +1,262 @@
+                         "Wrong reserved quants for the scanned package")
+
+        expected_mls = picking.move_line_ids.filtered(
+            lambda ml: ml.package_id == self.expected_package)
+        scanned_mls = picking.move_line_ids.filtered(
+            lambda ml: ml.package_id == self.scanned_package)
+
+        self.assertEqual(len(expected_mls), 0,
+                         "Got move lines related to the expected package")
+        self.assertEqual(len(scanned_mls), 1,
+                         "Don't have unique move line for the scanned package")
+        self.assertEqual(scanned_mls[0].package_id, self.scanned_package,
+                         "Move lines don't point to scanned package")
+        self.assertEqual(scanned_mls[0].result_package_id,
+                         self.scanned_package,
+                         "Move lines don't point to scanned package as result")
+
+    def test04_both_reserved(self):
+        """
+        Should successfully swap two packages:
+         - with same contents;
+         - in the same location;
+         - both scanned and expected packages are reserved.
+
+        In this case, the expected package shouldn't change
+        its state after update_picking is invoked.
+        """
+        self.create_quant(self.apple.id, self.test_location_01.id, 4,
+                          package_id=self.expected_package.id)
+        self.create_quant(self.apple.id, self.test_location_01.id, 4,
+                          package_id=self.scanned_package.id)
+        picking = self.create_picking(self.picking_type_pick,
+                                      products_info=self.pack_4apples_info,
+                                      confirm=True,
+                                      assign=True)
+        self.create_picking(self.picking_type_pick,
+                            products_info=self.pack_4apples_info,
+                            confirm=True,
+                            assign=True)
+
+        # checking pre-conditions
+        self.assertTrue(self.scanned_package.is_reserved())
+        self.assertTrue(self.expected_package.is_reserved())
+
+        # method under test
+        picking.update_picking(package_name=SCANNED_PACKAGE_NAME,
+                               expected_package_name=EXPECTED_PACKAGE_NAME)
+
+        # checking outcome
+
+        s_pack_quants = self.scanned_package._get_contained_quants()[0]
+
+        self.assertTrue(self.expected_package.is_reserved(),
+                        "Expected package is not reserved after swap")
+        self.assertTrue(self.scanned_package.is_reserved(),
+                        "Scanned package is not reserved after swap")
+        self.assertEqual(int(s_pack_quants.reserved_quantity), 4,
+                         "Wrong reserved quants for the scanned package")
+
+        scanned_mls = self.scanned_package.find_move_lines()
+
+        self.assertEqual(len(scanned_mls), 1,
```

Comment:
do we also want to assert that the picking_id of the scanned package move lines is picking?

Reply:
Yes, but that would be equal to the second picking instance, the one for the scanned package quant that is not currently assigned to any local.

Reply:
Which put a light on a bug!